File: Baekjoon_Java/프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class number {
    int x; 
    int y;
    public number(int x, int y){
        this.x = x;
        this.y = y;
    }
}
class Solution {
    static int [][] map;
    static boolean [][] visited;
    static int [] X = {1, -1, 0, 0};
    static int [] Y = {0, 0, 1, -1};
    public int solution(int[][] maps) {
        map = maps;
        visited = new boolean [map.length][map[0].length];
        visited[0][0] = true;
        bfs(0, 0);
        
        int answer = map[map.length-1][map[0].length-1];
        if(answer == 0 || answer == 1)
            return -1;
        else
            return answer;
    }
    static void bfs(int x, int y){
        Queue<number> queue = new LinkedList<>();
        queue.offer(new number(x, y));
        while(!queue.isEmpty()){
            number move = queue.poll();
            for(int i=0; i<4; i++){
                int newX = move.x + X[i];
                int newY = move.y + Y[i];
                if(newX >= 0 && newX < map.length && newY >= 0 && newY < map[0].length){
                    if(map[newX][newY]!=0 && !visited[newX][newY]){
                        visited[newX][newY] = true;
                        map[newX][newY] = map[move.x][move.y] + 1;
                        queue.offer(new number(newX, newY));
                    }
                }
            }
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    N, M = len(maps), len(maps[0])
    visited = [[False] * M for _ in range(N)]
    queue = deque([(0, 0)])
    visited[0][0] = True
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    while queue:
        x, y = queue.popleft()
        
        for d in range(4):
            nx, ny = x + dx[d], y + dy[d]
            
            if 0 <= nx < N and 0 <= ny < M and maps[nx][ny] == 1 and not visited[nx][ny]:
                visited[nx][ny] = True
                maps[nx][ny] = maps[x][y] + 1
                queue.append((nx, ny))
        
    if not visited[N-1][M-1]:
            return -1
    
    return maps[N-1][M-1]